clusterloader2/pkg/measurement/common/probes/probes.go

```diff
@@ -0,0 +1,326 @@
+		for _, sample := range samples {
+			quantile, err := strconv.ParseFloat(string(sample.Metric["quantile"]), 64)
+			if err != nil {
+				return nil, err
+			}
+			latency := time.Duration(float64(sample.Value) * float64(time.Second))
+			probe.latency.SetQuantile(quantile, latency)
+		}
+		if threshold, ok := thresholds[probeName]; ok {
+			if err := probe.latency.VerifyThreshold(threshold); err != nil {
+				err = errors.NewMetricViolationError(probe.name, err.Error())
+				klog.Errorf("%s: %v", p, err)
+			}
+		}
+		probeSummaries = append(probeSummaries, probe)
+	}
+	return &probesSummary{probeSummaries: probeSummaries}, err
+}
+
+func (p *probesMeasurement) createProbesObjects() error {
+	return p.framework.ApplyTemplatedManifests(manifestGlob, p.templateMapping)
+}
+
+func (p *probesMeasurement) waitTillProbesReady() error {
+	klog.Info("Waiting for Probes to become ready...")
+	return wait.Poll(checkProbesReadyInterval, checkProbesReadyTimeout, p.checkProbesReady)
+}
+
+func (p *probesMeasurement) checkProbesReady() (bool, error) {
+	serviceMonitors, err := p.framework.GetDynamicClients().GetClient().
+		Resource(serviceMonitorGvr).Namespace(probesNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		if client.IsRetryableAPIError(err) || client.IsRetryableNetError(err) {
+			err = nil // Retryable error, don't return it.
+		}
+		return false, err
+	}
+	// TODO(mm4tt): Using prometheus targets to check whether probes are up is a bit hacky.
+	//              Consider rewriting this to something more intuitive.
+	expectedTargets := p.replicasPerProbe * len(serviceMonitors.Items)
+	return prometheus.CheckTargetsReady(
+		p.framework.GetClientSets().GetClient(), isProbeTarget, expectedTargets)
+}
+
+func isProbeTarget(t prometheus.Target) bool {
+	return t.Labels["namespace"] == probesNamespace
+}
+
+func parseThresholds(params map[string]interface{}) (map[string]*measurementutil.LatencyMetric, error) {
+	thresholds := make(map[string]*measurementutil.LatencyMetric)
+	for name, thresholdVal := range params["thresholds"].(map[string]interface{}) {
+		threshold, err := time.ParseDuration(thresholdVal.(string))
+		if err != nil {
+			return nil, err
+		}
+		thresholds[name] = makeLatencyThreshold(threshold)
+	}
+	return thresholds, nil
+}
+
+func makeLatencyThreshold(threshold time.Duration) *measurementutil.LatencyMetric {
+	return &measurementutil.LatencyMetric{
+		Perc50: threshold,
+		Perc90: threshold,
+		Perc99: threshold,
+	}
+}
+
+func prepareQuery(queryTemplate string, startTime, endTime time.Time) string {
+	measurementDuration := endTime.Sub(startTime)
+	return fmt.Sprintf(queryTemplate, measurementutil.ToPrometheusTime(measurementDuration))
+}
+
+type probesSummary struct {
+	probeSummaries []*probeSummary
+}
+
+type probeSummary struct {
+	name    string
+	latency measurementutil.LatencyMetric
+}
+
+// SummaryName returns name of the summary.
+func (p *probesSummary) SummaryName() string {
+	return name
```

Comment:
This isn't very intuitive - I think that different probes are conceptually different things.
As a user, I would more expect to be separate summaries.

Why not simply have just probeSummary and return a separate one for every single probe?

Reply:
The motivation here was to group all the probes together and export in a single, unified form, so they can be easily (and generically) imported into PerfDash. My aim was to make adding a new probe as seamless as possible (in this case no changes required in the PerfDash at all). 

But I can change it and export each probe separately if you think that's better. PLMK.

Reply:
I think the goal is great as mentioned, and I think you're already roughly doing what I want:
- probe name is already in probeNameToPrometheusQueryTmpl and is propagate down to probeSummary
- you already have this "probeSummary" struct

I'm just suggesting to:
- remove `probesSummary` and leave only `probeSummary`
- the methods on probesSummary will now become methos on probeSummary
- return []probeSummary as the result (separate summary for every probe)

Does that make sense?

Reply:
Done.

Reply:
Partially - I thought it should be p.Name here.

Reply:
Doh! Thanks, fixed.